python_modules/dagster/dagster/_core/definitions/decorators/source_asset_decorator.py

```diff
@@ -0,0 +1,155 @@
+from typing import TYPE_CHECKING, Callable, Mapping, Optional, Sequence, Union, overload
+
+from typing_extensions import TypeAlias
+
+from dagster._core.definitions.events import AssetKey, AssetObservation, CoercibleToAssetKeyPrefix
+from dagster._core.definitions.logical_version import LogicalVersion
+from dagster._core.definitions.metadata import (
+    MetadataEntry,
+    MetadataUserInput,
+    PartitionMetadataEntry,
+)
+from dagster._core.definitions.partition import PartitionsDefinition
+from dagster._core.definitions.resource_definition import ResourceDefinition
+from dagster._core.definitions.source_asset import SourceAsset
+from dagster._core.errors import DagsterInvalidObservationError
+from dagster._core.execution.context.compute import OpExecutionContext
+from dagster._core.storage.io_manager import IOManagerDefinition
+
+from ..decorators.op_decorator import _Op
+
+if TYPE_CHECKING:
+    from dagster._core.execution.context.compute import SourceAssetObserveContext
+
+SourceAssetObserveFunction: TypeAlias = Union[
+    Callable[["SourceAssetObserveContext"], Union[MetadataUserInput, LogicalVersion]],
+    Callable[[], Union[MetadataUserInput, LogicalVersion]],
+]
+
+
+@overload
+def source_asset(observe_fn: SourceAssetObserveFunction) -> SourceAsset:
+    ...
+
+
+@overload
+def source_asset(
+    *,
+    name: Optional[str] = ...,
+    key_prefix: Optional[CoercibleToAssetKeyPrefix] = None,
+    metadata: Optional[MetadataUserInput] = None,
+    io_manager_key: Optional[str] = None,
+    io_manager_def: Optional[IOManagerDefinition] = None,
+    description: Optional[str] = None,
+    partitions_def: Optional[PartitionsDefinition] = None,
+    _metadata_entries: Optional[Sequence[Union[MetadataEntry, PartitionMetadataEntry]]] = None,
+    group_name: Optional[str] = None,
+    resource_defs: Optional[Mapping[str, ResourceDefinition]] = None,
+) -> "_SourceAsset":
+    ...
+
+
+def source_asset(
```

Comment:
what do you think about calling this something like `observeable_source_asset`?   I've thought about the idea of a different source asset decorator that contains a function that loads the source asset (which could then maybe be called `loadable_source_asset`).